File: Data Structures & Algorithms/reorder-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        int n=0;
        ListNode* temp=head;
        while(temp){
            n++;
            temp=temp->next;
        }
        if(n==1)return;
        int i=0;
        ListNode* temp2=head;
        while(temp2){
            if(i==n/2)break;
            temp2=temp2->next;
            i++;
        }
        ListNode* rev=NULL;
        ListNode* next=temp2->next;
        while(next!=NULL){
        next=temp2->next;
        temp2->next=rev;
        rev=temp2;
        if(next!=NULL)temp2=next;
        }
        ListNode*head1=head;
        while(temp2!=next){
            next=head1->next;
            if(temp2==next)break;
            head1->next=new ListNode(temp2->val);
            head1->next->next=next;
            head1=next;
            temp2=temp2->next;
        }
    }
};

File: Data Structures & Algorithms/add-two-numbers/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // ListNdoe x=l1;
        // ListNode y=l2;
        ListNode temp=new ListNode(0);
        ListNode res=temp;
        int digi=0;
        while(l1!=null&&l2!=null){
            int a=l1.val;
            int b=l2.val;
            int sum=a+b+digi;
            if(sum<10){
                digi=0;
                temp.next=new ListNode(sum);
            }
            else{
                digi=sum%10;
                sum=sum-digi;
                temp.next=new ListNode(digi);
                //temp=temp.next;
                //temp.next=new ListNode(sum);
                digi=sum/10;

            }
            temp=temp.next;
            l1=l1.next;
            l2=l2.next;
        }
        while(l1!=null){
            int a=l1.val;
            int sum=a+digi;
            if(sum<10){
                digi=0;
                temp.next=new ListNode(sum);
            }
            else{
                digi=sum%10;
                sum=sum-digi;
                temp.next=new ListNode(digi);
                //temp=temp.next;
                //temp.next=new ListNode(sum);
                digi=sum/10;

            }
            temp=temp.next;
            l1=l1.next;
        }
        while(l2!=null){
            int a=l2.val;
            int sum=a+digi;
            if(sum<10){
                digi=0;
                temp.next=new ListNode(sum);
            }
            else{
                digi=sum%10;
                sum=sum-digi;
                temp.next=new ListNode(digi);
                digi=sum/10;

            }
            temp=temp.next;
            l2=l2.next;
        }
        if(digi>0){
            temp.next=new ListNode(digi);
        }
        return res.next;
    }
}
